### PR TITLE
DEV/MNT: migrate to latest standards with pyproject.toml [LCLSPC-603]

### LINKED ISSUE #3: Cannot run pmgr.py script while developing
Relative imports in python do not work from the entry point script because the `__name__` attribute is set to `__main__` instead of the normal e.g. `pmgr.pmgr`

### LINKED ISSUE #19: Add requirements.txt and other PyPI necessities
Most of these should be on PyPI already:
```
    - pyqt >=5      -> PyQt5
    - pyca =3.1.1
    - epicscorelibs =7.0.4.99.1.2
    - pykerberos >=1.1.14
    - mysqlclient =1.3.12|>=2.0.3
    - docopt
    - pcdsutils
```

... except `pyca`?

### PR DESCRIPTION
**Pull request by klauer**
## Description
* Upgrading by way of https://github.com/pcdshub/pcds-python-migration-tools
* pmgr has been migrated to use setuptools-scm, replacing versioneer, as its version-string management tool of choice.
* pmgr has been migrated to use the modern ``pyproject.toml``, replacing ``setup.py`` and related files.
* Older language features and syntax found in the repository have been updated to Python 3.9+ standards by way of ``pyupgrade``.
* Sphinx 6.0 is now supported for documentation building. ``docs-versions-menu`` replaces ``doctr-versions-menu`` and ``doctr`` usage for documentation deployment on GitHub Actions.  The deployment key is now no longer required.
* `pyupgrade` was used to update annotations, though there might not have been many here
* `pycln` was used to clean up imports
* Many old files were removed
* Many old "readme" files were combined into the top-level project README
* Some logic errors were fixed
* uic-generated `ui.py` files were removed and are now dynamically loaded
* Some star imports were removed for qt5-related things

## Motivation and Context
* Migrating because we want to migrate our projects from Travis CI for one

Closes #3 
Closes #19 

## How Has This Been Tested?
* GitHub Actions

## Where Has This Been Documented?
This PR and Jira [LCLSPC-603](https://jira.slac.stanford.edu/browse/LCLSPC-603)

### CODE REVIEW

**Comment by klauer** [comment]
TLDR: I think this is probably not worth reviewing/merging at this point unless Mike indicates otherwise

A few things:
* I jumped in thinking this would be relatively small to fix up manually and regretted it
* There were a _ton_ of style fixes as you can see by the diff (which is understandable coming from a non-flake8, non-pep8, Python 2 repository). Some of the lint hid weird bugs that were wrapped with bare `try/except` and it's unclear if there's any side-effects this code was relying on 
* There are no tests to see if this is still doing what it should

**Comment by klauer** [comment]
Any opinions, @mcb64? Shelving it is probably the right approach, yeah?

**Comment by mcb64** [comment]
On the other hand, maybe we *should* just shelve this?!? :-)

Only kidding... I'll take a look.

**Comment by mcb64** [comment]
https://github.com/pcdshub/pmgr/blob/87e3e16343d41438f4f288a49cbc2c6fc093ae04/pmgr/CfgModel.py#L742-L745

This code was wrong, but we need to do:

```
try:
    todo = set(self.edits.keys())
except Exception:
    todo = set()
```

**Comment by mcb64** [comment]
https://github.com/pcdshub/pmgr/blob/87e3e16343d41438f4f288a49cbc2c6fc093ae04/pmgr/CfgModel.py#L936-L941

No need for todo here:
```
        if l is None:
            try:
                l = list(set(self.edits.keys()))
            except Exception:
                l = []
```

**Comment by mcb64** [comment]
https://github.com/pcdshub/pmgr/blob/87e3e16343d41438f4f288a49cbc2c6fc093ae04/pmgr/CfgModel.py#L992-L997

I don't think this function is actually used anywhere.  That said, these lines should be:
```
        try:
            self.root[self.root.index(old)] = new
        except Exception:
            pass
```

**Comment by mcb64** [comment]
*_ui.py should be deleted from the repo.  These are compiled .ui files.  If we don't want to bother pre-compiling them, we should change the rest of the program to use the .ui files directly.

**Comment by mcb64** [comment]
I think we can delete the migrate directory as well.

**Comment by klauer** [comment]
Thanks, @mcb64 and FYI you can click the big blue ➕  when mousing over a line on the diff to add specific line comments/suggestions: 
image

**Comment by klauer** [comment]
OK, I've addressed the above. 

I don't really know how to use parameter manager if I'm being honest, though, so I've only managed to confirm that it still pops up a dialog appropriately in our standard pcds-5.6.0 environment:

image

File: pmgr/OBSOLETE/utilsPlus.py [hunk 1]
```diff
@@ -227,76 +245,134 @@ def getFieldDict(pmgr, PV, fields):
...
+    somewhere along the pipeline. It will return a corrected configuration
     dictionary if the SN length is wrong, and return the same one if it is fine.
     """
 
     # Make sure it is a string
     cfgDict["FLD_SN"] = str(cfgDict["FLD_SN"])
-    
-    SN = cfgDict["FLD_SN"]
+
+    # SN = cfgDict["FLD_SN"]
```
Comment (mcb64) [inline pmgr/OBSOLETE/utilsPlus.py]: Yeah, just delete this.

File: pmgr/harvester.py [hunk 2]
```diff
@@ -3,28 +3,29 @@
 # A bizarre combination of code from IocManager, the parent/child IOC compilation process, and
```
Comment (mcb64) [inline pmgr/harvester.py:3]: I think this first line sums it up.

I'm not convinced this would work any more, and even if it did, our database is mostly populated at this point.

Let's kill this.

Comment (klauer) [reply pmgr/harvester.py]: Done in fc6f8b9

File: pmgr/pmgrAPI.py [hunk 3]
```diff
@@ -72,11 +73,13 @@ def _fixmutex(d, mutex):
             The original configuration dictionary with the derived fields
             removed.
         """
+        # NOTE: fixed as this cannot be a staticmethod since it refers to 'self.pm'
```
Comment (mcb64) [inline pmgr/pmgrAPI.py:76]: Yeah, kill these comments, the fixes are good.

**Review by mcb64 — APPROVED** [review]
Yeah, that's a lot, but I think it's all good.

File: conda-recipe/meta.yaml [hunk 4]
```diff
@@ -1,41 +1,44 @@
...
-    - pcdsutils
+  - python >=3.9
+  - pyqt >=5
+  - pyca =3.2.0
+  - epicscorelibs =7.0.6.99.2.0
+  - pykerberos >=1.1.14
+  - mysqlclient =1.3.12|>=2.0.3
+  - docopt
+  - epicscorelibs =7.0.4.99.1.2
+  - mysqlclient =1.3.12|>=2.0.3
```
Comment (klauer) [inline conda-recipe/meta.yaml:31]:
```suggestion
```

Rebase failure. Correct deps are up above.

**Review by tangkong — APPROVED** [review]
I gave it a quick once over and most things seemed reasonable.  Albeit I got to look at it post-mike-edits.  

The tests fail to find the `R2.1.1` tag even though it exists on github.  Some weird versioning shenanigans going on here certaintly

**Comment by klauer** [comment]
I _think_ setuptools-scm expects to see `vX.Y.Z`-style version numbers. I'm not sure if we can reconfigure it to use the LCLS/EPICS-specific `Rx.y.z`-style - but it's probably worth looking into

**Comment by klauer** [comment]
OK, I've gotten at least the package building and the requirements specified.

There are no tests here, and not all of our dependencies are on PyPI (pyca) to even do an import test.

This is _good enough_. I'm going to merge so we can close that Jira migration issue and move on.
